### ISSUE TITLE
image pull might silently ignore errors

### ISSUE THREAD

**Issue by cacheflux** [containrrr/watchtower#344]
I've been reading through the source and I'm wondering if I found an edgecase in which an image pull can silently fail. There's a chance I'm missing something here, please let me know if that's the case.

https://github.com/containrrr/watchtower/blob/9eca883f17669bc93e0f1dc4b9be95b7f3684656/container/client.go#L237-L251

The comment in line 244 notes that we need to stream the reply, which suggests the pull started in line 237 is still ongoing and err being nil in line 238 doesn't mean the pull was successfully completed.

In line 245 we assign a new value to err after the pull has finished, yet in lin 248 we reassigning a new value to err, which means the previous value is never accessed.

In that case the code would continue to inspect the old image as if pulling had been disabled and conclude that no update is available.

**Comment by simonaronsson** [comment]
Thank you for the report!

I'm inclined to agree with you. I think this might require some testing though as we're not really interested in reading the response but rather do it to avoid aborting prematurely. Not really sure whether err in ReadAll only returns errors related to reading or if it also takes the pull status into account.

Feel free to investigate it further and if possible even provide a PR and i'll make sure to review it. 👍

**Comment by cacheflux** [comment]
Correct, checking err would only tell us if `ReadAll` errored (which shouldn't happen unless the docker daemon died or we're using a remote docker daemon). To make sure we log pull failures correctly we need to parse the stream we're discarding and check for errors. The response looks like this:
```
{"status":"Extracting","progressDetail":{"current":458752,"total":45339350},"progress":"[\u003e                                                  ]  458.8kB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":3670016,"total":45339350},"progress":"[====\u003e                                              ]   3.67MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":7340032,"total":45339350},"progress":"[========\u003e                                          ]   7.34MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":11010048,"total":45339350},"progress":"[============\u003e                                      ]  11.01MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":14680064,"total":45339350},"progress":"[================\u003e                                  ]  14.68MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":18350080,"total":45339350},"progress":"[====================\u003e                              ]  18.35MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":21102592,"total":45339350},"progress":"[=======================\u003e                           ]   21.1MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":28442624,"total":45339350},"progress":"[===============================\u003e                   ]  28.44MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":33030144,"total":45339350},"progress":"[====================================\u003e              ]  33.03MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":36241408,"total":45339350},"progress":"[=======================================\u003e           ]  36.24MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":39452672,"total":45339350},"progress":"[===========================================\u003e       ]  39.45MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":42663936,"total":45339350},"progress":"[===============================================\u003e   ]  42.66MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":44498944,"total":45339350},"progress":"[=================================================\u003e ]   44.5MB/45.34MB","id":"6f2f362378c5"}
{"status":"Extracting","progressDetail":{"current":45339350,"total":45339350},"progress":"[==================================================\u003e]  45.34MB/45.34MB","id":"6f2f362378c5"}
{"status":"Pull complete","progressDetail":{},"id":"6f2f362378c5"}
{"status":"Digest: sha256:118cf8f3557e1ea766c02f36f05f6ac3e63628427ea8965fb861be904ec35a6f"}
{"status":"Status: Downloaded newer image for debian:latest"}
```

In case of an error:
```
{"status":"Pulling from library/debian","id":"latest"}
{"errorDetail":{"message":"Get https://registry-1.docker.io/v2/library/debian/manifests/sha256:0faf75dd7375b860af2589808c5b26af60a90c0d50a8a4d4d2f32974d7aff27d: read tcp 192.168.1.3:44576-\u003e34.199.40.84:443: read: connection reset by peer"},"error":"Get https://registry-1.docker.io/v2/library/debian/manifests/sha256:0faf75dd7375b860af2589808c5b26af60a90c0d50a8a4d4d2f32974d7aff27d: read tcp 192.168.1.3:44576-\u003e34.199.40.84:443: read: connection reset by peer"}
```

**Comment by simonaronsson** [comment]
Thanks! You seem to have both repro and a solution, so if you don't mind it would be awesome with a PR correcting the issue. I'll make sure to review it and merge it as soon as possible. 🙏

**Comment by simonaronsson** [comment]
thanks stalebot, but please keep this open

**Comment by simonaronsson** [comment]
Keep this open, thanks

**Comment by simonaronsson** [comment]
Solved in d1abce8
